src/main/java/hudson/plugins/plot/PlotBuildAction.java

```diff
@@ -0,0 +1,42 @@
+package hudson.plugins.plot;
+
+import hudson.model.Action;
+import hudson.model.InvisibleAction;
+import hudson.model.Run;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import jenkins.tasks.SimpleBuildStep;
+import org.kohsuke.stapler.StaplerProxy;
+import org.kohsuke.stapler.export.ExportedBean;
+
+@ExportedBean
+public class PlotBuildAction extends InvisibleAction implements StaplerProxy, SimpleBuildStep.LastBuildAction {
+
+    private Run<?, ?> run;
+    private List<Plot> plots;
+
+    PlotBuildAction(Run<?, ?> run, List<Plot> plots) {
+        this.run = run;
+        this.plots = plots;
+    }
+
+    @Override
+    public synchronized Collection<? extends Action> getProjectActions() {
+        return Collections.<Action>singleton(new PlotAction(run.getParent(), plots));
+    }
+
+    @Override
+    public Object getTarget() {
+        return null;
+    }
+
+    synchronized void addPlots(List<Plot> plots) {
```

Comment:
Serialization may not invoke getters & setters by the way -- depending on the implementation and annotations it may be doing direct field access using reflection -- which means the synchronization won't necessarily project vs. ConcurrentModificationExceptions.  See my comment on the commit for a strategy to protect against that.

Reply:
Though I'm a bit at a loss why serialization is being invoked *during* mutation rather than before/after?  That might be a bug somewhere.

Reply:
Not sure what's better in this case "synchronized" or "array copy". It works in both ways. Let's leave it as it is if no other concerns appear.

Reply:
Well, that user still seemed to be having issues after the use of synchronized so I'm inclined to think that one needs a slightly deeper dive (and maybe copy-on-write).

Reply:
I'm not against copy-on-write approach, I just feel that it will be implemented again "blindly", as issue was reported by @dolfinsbizou but we're not sure what causes the issue and no STR's.